zcash_client_sqlite/src/wallet.rs

```diff
@@ -200,10 +197,47 @@ pub(crate) fn get_current_address<P: consensus::Parameters>(
     .transpose()
 }
 
+/// Adds the given address and diversifier index to the addresses table.
+///
+/// Returns the database row for the newly-inserted address.
+pub(crate) fn insert_address<P: consensus::Parameters>(
+    conn: &rusqlite::Connection,
+    params: &P,
+    account: AccountId,
+    mut diversifier_index: DiversifierIndex,
+    address: &UnifiedAddress,
+) -> Result<(), rusqlite::Error> {
+    let mut stmt = conn.prepare_cached(
+        "INSERT INTO addresses (
+                    account,
+                    diversifier_index_be,
+                    address,
+                    cached_transparent_receiver_address
+                )
+                VALUES (
+                    :account,
+                    :diversifier_index_be,
+                    :address,
+                    :cached_transparent_receiver_address
+                )",
+    )?;
+
+    // the diversifier index is stored in big-endian order to allow sorting
+    diversifier_index.0.reverse();
+    stmt.execute(named_params![
+        ":account": &u32::from(account),
+        ":diversifier_index_be": &&diversifier_index.0[..],
+        ":address": &address.encode(params),
+        ":cached_transparent_receiver_address": &address.transparent().map(|r| r.encode(params)),
+    ])?;
+
+    Ok(())
+}
+
 #[cfg(feature = "transparent-inputs")]
 pub(crate) fn get_transparent_receivers<P: consensus::Parameters>(
+    conn: &rusqlite::Connection,
     params: &P,
-    conn: &Connection,
     account: AccountId,
 ) -> Result<HashMap<TransparentAddress, AddressMetadata>, SqliteClientError> {
```

Comment:
Nit for preexisting code: handling of invalid addresses is duplicated between here and `get_current_address`.